### PR TITLE
Initial derivation remote cache related metrics: check, download and upload time.

### PR DESCRIPTION
**Pull request by restingbull**
* Introduces building block datum, LocalActions, to make it easier to reason about actions.
* Introduces RemoteCacheMetrics as basic reporting datum for cache performance.

### CODE REVIEW

**Comment by saraadams** [comment]
@restingbull I haven't had a look at this, as I wanted to give you time to address Louis' comments first. Let me know if you are waiting for my review first.

**Comment by restingbull** [comment]
Been a bit busy to address all the comments. Will get to it this week.

**Comment by afrueda97** [comment]
I've added the suggestions to the pull request @lrgoldstein , what do you think?

File: analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/LocalActions.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+  public boolean isEmpty() {
+    return actions.isEmpty();
+  }
+
+  @Override
+  public String getEmptyReason() {
+    if(isEmpty()) {
+      return "No local actions executed";
+    } else {
+      return "This is non-empty.";
```
Comment (lrgoldstein) [inline analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/LocalActions.java:77]: Other providers just return `null`. Suggest being consistent.

File: analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/LocalActionsDataProvider.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+public class LocalActionsDataProvider extends DataProvider {
+
+  @Override
+  public List<DatumSupplierSpecification<?>> getSuppliers() {
+    return List.of(
+        DatumSupplierSpecification.of(LocalActions.class, DatumSupplier.memoized(this::derive)));
+  }
+
+
```
Comment (lrgoldstein) [inline analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/LocalActionsDataProvider.java:39]: Nit: too much space here.

File: analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetricsDataProvider.java [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        metrics.parallelStream().map(d -> d.download).reduce(Duration::plus).orElse(Duration.ZERO),
+        metrics.parallelStream().map(d -> d.upload).reduce(Duration::plus).orElse(Duration.ZERO),
+        (metrics.parallelStream()
+                    .map(d -> d.check.isZero() || d.download.isZero() ? 0 : 1f)
+                    .reduce(0f, Float::sum)
+                / metrics.size())
+            * 100);
+  }
+
+  RemoteCacheData coalesce(LocalAction action) {
```
[the remark this answers is not in the document]
Comment (restingbull) [reply analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetricsDataProvider.java:50]: co·a·lesce
/ˌkōəˈles/
verb
come together to form one mass or whole.

I suppose we could use aggregate or summarize?

File: analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetricsDataProvider.java [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    return events.stream()
+        .filter(e -> category.equals(e.category))
+        .map(e -> e.duration)
+        .findFirst()
+        .orElse(Duration.ZERO);
+  }
+
+  static class RemoteCacheData {
+
+    public final String name;
```
[the remark this answers is not in the document]
Comment (restingbull) [reply analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetricsDataProvider.java:68]: Future work, really. Was intending to add some actual useful data to the metrics.

File: analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  @Override
+  public boolean isEmpty() {
+    return totalCacheCheck.isZero() && totalDownloadOutputs.isZero();
+  }
+
+  @Override
+  public String getEmptyReason() {
+    return "No Remote Cache checks or downloads";
```
[the remark this answers is not in the document]
Comment (restingbull) [reply analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java:33]: That's a bit of an unexpected api.

File: analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+  }
+
+  @Override
+  public String getDescription() {
+    return "Remote cache usage.";
```
[the remark this answers is not in the document]
Comment (restingbull) [reply analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java]: Added more description.

File: analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    return "RemoteCacheMetrics{" +
+        "totalCacheCheck=" + totalCacheCheck +
+        ", totalDownloadOutputs=" + totalDownloadOutputs +
+        ", totalUploadOutputs=" + totalUploadOutputs +
+        ", percentCached=" + percentCached +
+        '}';
+  }
+
+  @Override
+  public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (restingbull) [reply analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java:60]: Tests.

**Comment by restingbull** [comment]
> > I've added the suggestions to the pull request @lrgoldstein , what do you think?
> 
> I've gone through and resolved the items that you addressed and added a couple... there were also a handful from before that weren't addressed (be sure to "Load" the "hidden conversations" in the middle).
> 
> Edit: Also, you have to [sign all your commits](https://github.com/EngFlow/bazel_invocation_analyzer/blob/main/CONTRIBUTING.md#process).

You have to do a bit more that just sign them. I have `commit.gpgsign true` (hence I have Screen Shot 2023-03-01 at 8 29 34 PM) -- which means I can prove I made the commit...

You have to a more explicit `--sign-off`, which means you stand by it (I guess?). Thanks, git, for not making this confusing.

File: analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (lrgoldstein) [reply analyzer/java/com/engflow/bazel/invocation/analyzer/dataproviders/RemoteCacheMetrics.java]: You would expect it to have an Empty Reason if it's not Empty? It's typically determined dynamically, as you could run into a number of different problems while producing a datum depending on the complexity of the calculation.

**Review by lrgoldstein — APPROVED** [review]

**Pull request merged**